[TF-TRT] Introduce `_remove_native_segments` function to remove native segments in TF-TRT Graph

pavanimajety opened the pull request:
This function parses through the TF-TRT graph and deletes native segment nodes constructed for falling back. Changes have also been made to `TRTEngineOp` to recognize when the loaded graph doesn't have `_native_segment` nodes and disallow loading the graph associated with the native segments. 

To do: When native segment execution is disabled, we don't generate the native_segment nodes in the first place.


@tfeher and @DEKHTIARJonathan  please review.

tfeher commented on tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -721,7 +730,11 @@ void TRTEngineOp::ExecuteCalibration(OpKernelContext* ctx,
+  if(!native_segment_absent_) {
+    ExecuteNativeSegment(ctx, async_helper);
+  } else{
+    LOG(ERROR)<< "Re-run calibration with original saved model graph, native segment not found";
```
I would change the wording. We use the converter object to run calibration, and that always has the original graph. Calibration is not something that we can re-run with a different graph: we change the saved model only when we save the converted graph, but we cannot save the model without calibration.

```suggestion
    LOG(ERROR)<< "Calibration requires native segment, but it is not available in in the graph";
```

tfeher commented on tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -869,7 +882,11 @@ void TRTEngineOp::ComputeAsync(OpKernelContext* ctx,
+      if(!native_segment_absent_) {
+        ExecuteNativeSegment(ctx, async_helper);
+      } else{
+        LOG(ERROR)<< "Re-run calibration with original saved model graph, native segment not found";
```
```suggestion
        LOG(ERROR)<< "Expected to use native segment during profile generation, but native segment not found";
```

tfeher commented on tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -954,14 +971,19 @@ void TRTEngineOp::ComputeAsync(OpKernelContext* ctx,
   if (!may_execute_native_segment()) {
```
Maybe add `native_segment_absent_` to the definition of `may_execute_native_segment`, so that it is utilized here as well.

tfeher commented on tensorflow/python/compiler/tensorrt/trt_convert_test.py:
```diff
@@ -702,6 +702,34 @@ def _InputFn():
+    converter.summary()
+    converter.build(_input_fn)
+    # Load and verify the converted model.
+    graph_func = trt_convert._remove_native_segment(graph_func)
```
You probably want to run the function to test it. Additionally, it would be good to save & load the model https://github.com/tensorflow/tensorflow/blob/4f8b87739e3f3c4aee0e36714e16856227c83e9a/tensorflow/python/compiler/tensorrt/trt_convert_test.py#L1027-L1032
and infer the loaded model https://github.com/tensorflow/tensorflow/blob/4f8b87739e3f3c4aee0e36714e16856227c83e9a/tensorflow/python/compiler/tensorrt/trt_convert_test.py#L1052

tfeher commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -842,11 +842,47 @@ def _restore_collections(dest_graph, src_meta_graph_def, collection_keys):
-    return gen_trt_ops.create_trt_resource_handle(resource_name=name)
+    return gen_trt_ops.create_trt_resource_handle(resource_name=name) 
+
+def _remove_native_segment(input_func):
```
Shall we have an extra arg for `save` that would turn this on?

pavanimajety replied:
Added `remove_native_segments` flag.

tfeher reviewed (changes requested):
Hi Pavani! Thanks for this PR! It looks great overall, I have just a few smaller comments.

DEKHTIARJonathan commented:
@pavanimajety @tfeher instead of introducing an API, could we use for now: `is_experimental_feature_activated` defined here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/compiler/tensorrt/utils.py#L79.

I would like to avoid introducing an API and later changing it because it's not a fortunate design.

For now, let's keep it simple and avoid any API change. Once the feature is complete and properly tested, we'll think about what's the best way(s) to expose the feature to the end-user.

@pavanimajety what do you think about `TF_TRT_EXPERIMENTAL_FEATURES=remove_native_segments` ? If you have an idea that you prefer, it's fine with me. It doesn't really matter :)

tfeher commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -842,11 +842,48 @@ def _restore_collections(dest_graph, src_meta_graph_def, collection_keys):
+  Args:
+    input_func: provide the concrete function with native segment nodes. 
+     The transformed output func will not contain any native segment nodes.
+     All the TRTEngineOp references will be modified to a string "deleted"
+     to automatically identify that the associated native segment is deleted.
```
This is not true anymore, right? It seems that you are deleting the `segment_func` attribute.  
IIUC, this means that the `segment_func` attribute will have a default value default value, which has an empty name.

pavanimajety replied:
That is correct, I fixed it.

tfeher commented on tensorflow/python/compiler/tensorrt/trt_convert_test.py:
```diff
@@ -702,6 +702,44 @@ def _InputFn():
+    converter.build(_input_fn)
+    # Load and verify the reduced converted model.
+    output_saved_model_dir2 = self.mkdtemp()
+    os.environ["TF_TRT_EXPERIMENTAL_FEATURES"] = "remove_native_segments"
+    converter.save(output_saved_model_dir2)
```
We have a context manager for this
```suggestion
    with test_utils.experimental_feature_scope("remove_native_segments"):
      converter.save(output_saved_model_dir2)
```

tfeher reviewed (changes requested):
Thanks @pavanimajety for the updates! I just have two small comments

pavanimajety committed:
[TF-TRT] Provide functions to remove native segments in TF-TRT Graphs

Add unit test and fix logging message
Address Tamas' feedback, introduce  experimental env variable to save method. Use experimental feature test method, update method description.

tfeher reviewed (approved):
Thanks Pavani for the update, the PR looks good to me!
